components/cli/pkg/registry/credentials/credentials.go

```diff
@@ -0,0 +1,185 @@
+			util.ExitWithErrorMessage("Error while establishing the service", err)
+		}
+	}()
+
+	err := util.OpenBrowser(gitHubAuthUrl)
+	if err != nil {
+		fmt.Printf("Could not resolve the given url %s. Started to operate in the headless "+
+			"mode\n", gitHubAuthUrl)
+		return FromTerminal(username)
+	}
+	// Setting up a timeout
+	go func() {
+		time.Sleep(5 * time.Minute)
+		timeout <- true
+	}()
+	// Wait for a code, or timeout
+	select {
+	case <-ch:
+	case <-timeout:
+		util.ExitWithErrorMessage("Failed to authenticate", errors.
+			New("time out. Did not receive any code"))
+	}
+	token := getTokenFromCode(code, codeReceiverPort, auth)
+	username, accessToken := getUsernameAndTokenFromJWT(token)
+	return username, accessToken, nil
+}
+
+// FromTerminal is to allow this login flow to work in headless mode
+func FromTerminal(username string) (string, string, error) {
+	var password string
+	if username == "" {
+		fmt.Print("Enter username: ")
+		_, err := fmt.Scanln(&username)
+		if err != nil {
+			util.ExitWithErrorMessage("Error reading the input username", err)
+		}
+	}
+	fmt.Print("Enter password/token: ")
+	bytePassword, err := terminal.ReadPassword(0)
+	if err != nil {
+		util.ExitWithErrorMessage("Error reading the input token", err)
+	}
+	password = strings.TrimSpace(string(bytePassword))
+	fmt.Println()
+	return username, password, nil
+}
+
+// getUsernameAndTokenFromJWT returns the extracted subject from the JWT
+func getUsernameAndTokenFromJWT(token string) (string, string) {
+	var result map[string]interface{}
+	err := json.Unmarshal([]byte(token), &result)
+	if err != nil {
+		util.ExitWithErrorMessage("Error while unmarshal the id_token", err)
+	}
+	idToken, ok := (result["id_token"]).(string)
+	accessToken, ok := (result["access_token"]).(string)
+	if !ok {
+		util.ExitWithErrorMessage("Error while retrieving the access token", err)
+	}
+	jwtToken, _ := jwt.Parse(idToken, nil)
+	claims := jwtToken.Claims.(jwt.MapClaims)
+	sub, ok := claims["sub"].(string)
+	if !ok {
+		util.ExitWithErrorMessage("Error in casting the subject", err)
+	}
+	return sub, accessToken
+}
+
+// getTokenFromCode returns the JWT from the auth code provided
+func getTokenFromCode(code string, port int, auth config.AuthConf) string {
+	tokenUrl := "https://" + auth.IsHost + ":" + strconv.Itoa(auth.IsPort) + "/oauth2/token"
+	responseBody := "client_id=" + auth.SpClientId + "&grant_type=authorization_code&code=" + code +
+		"&redirect_uri=http://localhost:" + strconv.Itoa(port) + auth.CallBackContextPath
+	body := strings.NewReader(responseBody)
+	// Token request
+	req, err := http.NewRequest("POST", tokenUrl, body)
+	if err != nil {
+		util.ExitWithErrorMessage("Error while creating the code receiving request", err)
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded;charset=UTF-8")
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
```

Comment:
Shall we add a TODO comment (to remove this later) here ? This is something that shouldn't go to production in the end.